File: ballast-client/dotnet/src/signalr/services/SignalRClientApplicationEventEmitter.cs
using System.Threading.Tasks;
using Ballast.Client.SignalR.Services;
using Ballast.Core.Application.Events;
using Ballast.Core.Application.Services;
using Ballast.Core.Messaging;
using Microsoft.AspNetCore.SignalR.Client;

namespace Ballast.Client.SignalR.Services 
{
    public class SignalRClientApplicationEventEmitter : SignalRClientServiceBase, IApplicationEventEmitter
    {

        private readonly IEventBus _eventBus;
        private bool _isEnabled;

        public SignalRClientApplicationEventEmitter(ISignalRClientOptions options, IEventBus eventBus) : base(options)
        {
            _eventBus = eventBus;
            _isEnabled = false;
        }

        protected override string HubName => "eventhub";

        public bool IsEnabled => _isEnabled;

        public Task StartAsync()
        {
            _isEnabled = true;
            return Task.CompletedTask;
        }

        public Task StopAsync()
        {
            _isEnabled = false;
            return Task.CompletedTask;
        }

        protected override void AfterSubscribe(HubConnection hubConnection)
        {
            hubConnection.On<IApplicationEvent>("IApplicationEvent", OnApplicationEvent);
        }

        protected override void BeforeUnsubscribe(HubConnection hubConnection)
        {
            hubConnection.Remove("IApplicationEvent");
        }

        private void OnApplicationEvent(IApplicationEvent evt)
        {
            if (!_isEnabled)
                return;
            _ = _eventBus.PublishAsync(evt); // Fire and forget
        }

    }
}

File: ballast-client/npm/src/signalr/services/signalr-client-application-event-emitter.ts
import { HubConnection } from "@aspnet/signalr";
import { IApplicationEvent, IApplicationEventEmitter, IEventBus, TYPES as BallastCore } from "ballast-core";
import { inject, injectable } from "inversify";
import { TYPES as BallastClient } from "../../dependency-injection/types";
import { ISignalRClientOptions } from "../signalr-client-options";
import { SignalRClientServiceBase } from "../signalr-client-service-base";

@injectable()
export class SignalRClientApplicationEventEmitter extends SignalRClientServiceBase implements IApplicationEventEmitter {

    public static readonly hubName: string = "eventhub";
    public get hubName() {
        return SignalRClientApplicationEventEmitter.hubName;
    }

    private readonly _eventBus: IEventBus;
    private _isEnabled: boolean;

    public constructor(
        @inject(BallastClient.SignalR.ISignalRClientOptions) serviceOptions: ISignalRClientOptions,
        @inject(BallastCore.Messaging.IEventBus) eventBus: IEventBus
    ) {
        super(serviceOptions);
        this._eventBus = eventBus;
        this._isEnabled = false;
        this.rebindHandlers();
    }

    public get isEnabled(): boolean {
        return this._isEnabled;
    }

    public start(): void {
        this._isEnabled = true;
    }

    public stop(): void {
        this._isEnabled = false;
    }

    protected rebindHandlers() {
        this.onApplicationEvent = this.onApplicationEvent.bind(this);
    }

    protected afterSubscribe(hubConnection: HubConnection) {
        hubConnection.on("IApplicationEvent", this.onApplicationEvent);
    }

    protected beforeUnsubscribe(hubConnection: HubConnection) {
        hubConnection.off("IApplicationEvent");
    }
    
    private onApplicationEvent(evt: IApplicationEvent) {
        if (!this._isEnabled) {
            return Promise.resolve();
        }
        this._eventBus.publishAsync(evt); // Fire and forget
    }

}